source/common/json/json_serializer.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/buffer/buffer.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "fmt/format.h"
+#include "json_internal.h"
+
+namespace Envoy {
+namespace Json {
+
+using BufferWriter = Envoy::Buffer::Instance;
+
+class StringWriter {
+public:
+  StringWriter(size_t initial_buffer_size = 2048) { buffer_.reserve(initial_buffer_size); }
+
+  void addFragments(absl::Span<const absl::string_view> fragments) {
+    for (absl::string_view fragment : fragments) {
+      buffer_.append(fragment.data(), fragment.size());
+    }
+  }
+  std::string buffer_;
+};
+
+/**
+ * Helper class to sanitize keys, values, and delimiters to short JSON pieces
+ * (strings).
+ * NOTE: Use this class carefully. YOURSELF should make sure the JSON pieces
+ * are sanitized and constructed correctly.
+ * NOTE: Although this class could but is not designed to construct a complete
+ * JSON output. It is designed to sanitize and construct partial JSON pieces
+ * on demand.
+ * NOTE: If a complete JSON output is needed, using Envoy::Json::Streamer first
+ * except YOU KNOW WHAT YOU ARE DOING.
+ */
```

Comment:
this is completely redundant in functionality with json_streamer.h and I think we should figure out how to merge them so they work well in buffer and string contexts.

At a very minimum, we should write a  comment in both classes

```Note: JsonSerializer does exactly the same thing as JsonStreamer, except using strings. Choose which to use based on context.```

you'd put that here and flip it in json_streamer.h.

I reallyh hate duplicating this functilonality though. It makes it very hard to understand what to use for someone else coming in.

Reply:
I actually got a Idea. See, basically, the serializer only includes some very basic methods that used to sanitize and write data to buffer, like addString().
And these methods are duplicate with streamer's.

We can let the streamer to inherit from serializer or make the serializer as a data member of streamer.

The current serializer is a template class that could take a templated `Writer` to outputs sanitized data.

The `Writer` class only need to provides a addFragmemts() method. (Yeah, addFragmemts method. The Buffer::Instance could be used as `Writer` directly.)

So, it would be pretty easy to merge them once we can make the decision to remove exist `sanitize()`.

Reply:
some refactoring strategy where the code is shared makes sense.

I was thinking of templatizing JsonStreamer and its nested classes on the buffer type, and make a wrapper for std::string that just creates addFragments() and add() wrappers so the call-site is the same.

THat would make me feel a lot more confident than re-creating all the json serialization logic.

Reply:
Haha, see the `StringWriter`. I think I have created this wrapper.

Reply:
What I was trying to do was centralize the json serializing logic in one place and use it both in this context and when streaming a deeper structured output. See https://github.com/envoyproxy/envoy/pull/35647 for one way to go about this.

What you have done still is to duplicate the serialization logic, which I think adds technical debt and risk for having inconsistencies and bugs.

Another approach is to factor out the serialization logic from json_streamer -- that would be a little more intrusive than the templated approach I proposed in #35647 but might be a little cleaner.

Reply:
I prefer to split the basic serialization method out from the Streamer. It would be cleaner and won't bring much complexity.